pkg/guestagent/iptables/iptables.go

```diff
@@ -0,0 +1,111 @@
+package iptables
+
+import (
+	"bytes"
+	"errors"
+	"net"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type Entry struct {
+	IP   net.IP
+	Port int
+}
+
+// This regex can detect a line in the iptables added by portmap to do the
+// forwarding. The following two are examples of lines (notice that one has the
+// destination IP and the other does not):
+//    -A CNI-DN-2e2f8d5b91929ef9fc152 -d 127.0.0.1/32 -p tcp -m tcp --dport 8081 -j DNAT --to-destination 10.4.0.7:80
+//    -A CNI-DN-04579c7bb67f4c3f6cca0 -p tcp -m tcp --dport 8082 -j DNAT --to-destination 10.4.0.10:80
+// The -A on the front is to amend the rule that was already created. portmap
+// ensures the rule is created before creating this line so it is always -A.
+// CNI-DN- is the prefix used for rule for an individual container.
+// -d is followed by the IP address. The regular expression looks for a valid
+// ipv4 IP address. We need to detect this IP.
+// --dport is the destination port. We need to detect this port
+// -j DNAT this tells us it's the line doing the port forwarding.
+var findPortRegex = regexp.MustCompile(`-A\s+CNI-DN-\w*\s+(?:-d ((?:\b25[0-5]|\b2[0-4][0-9]|\b[01]?[0-9][0-9]?)(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}))?(?:/32\s+)?-p .*--dport (\d+) -j DNAT`)
```

Comment:
I don't think we should limit the regex to CNI rules.
We should cover non-CNI rules such as Docker libnetwork rules with `userland-proxy: false`.

We can just overestimate the set of the listening ports, and probably try `connect()`-ing to verify whether the port is actually open (for TCP ports).

Reply:
I tested lima with Docker and `userland-proxy: false`. The port forwarding with the latest release of lima and this patch works without any changes. The iptables rule for forwarding is listed and reads...

```
-A DOCKER -p tcp -m tcp --dport 8081 -j DNAT --to-destination 172.17.0.2:80
```

When the userland proxy is in place the iptables rule is different...

```
-A DOCKER -d 127.0.0.1/32 ! -i docker0 -p tcp -m tcp --dport 8081 -j DNAT --to-destination 172.17.0.2:80
```

Docker (I think moby) is doing something to make the ports work. For example, running `netstat -plunt` will show the forwarded ports for docker but not containerd (running as root).

Do we need to support Docker if it already works?

Note, I tested docker running rootless and as root.

Reply:
I updated the iptables check to verify TCP ports are open before returning them.